File: backend/src/specforge/discovery_options.py
from __future__ import annotations

DISCOVERY_CUSTOM_OPTION_LABEL = "其他（请说明）"
MIN_DISCOVERY_PRESET_OPTIONS = 1


def normalize_discovery_options(options: list[str]) -> list[str]:
    """Ensure ask-turn options are non-empty with a custom-input choice last."""
    presets = [item.strip() for item in options if item and item.strip()]
    presets = [item for item in presets if item != DISCOVERY_CUSTOM_OPTION_LABEL]
    if len(presets) < MIN_DISCOVERY_PRESET_OPTIONS:
        raise ValueError(
            f"discovery ask requires at least {MIN_DISCOVERY_PRESET_OPTIONS} preset option "
            f"plus '{DISCOVERY_CUSTOM_OPTION_LABEL}'"
        )
    return [*presets, DISCOVERY_CUSTOM_OPTION_LABEL]


def is_discovery_custom_option(label: str) -> bool:
    return label.strip() == DISCOVERY_CUSTOM_OPTION_LABEL

File: frontend/src/shared/lib/discoveryOptions.ts
export const DISCOVERY_CUSTOM_OPTION_LABEL = '其他（请说明）'

export function splitDiscoveryOptions(options: string[]) {
  if (options.length === 0) {
    return { presets: [] as string[], customLabel: DISCOVERY_CUSTOM_OPTION_LABEL }
  }
  const last = options[options.length - 1]
  if (last === DISCOVERY_CUSTOM_OPTION_LABEL) {
    return { presets: options.slice(0, -1), customLabel: last }
  }
  return { presets: options, customLabel: DISCOVERY_CUSTOM_OPTION_LABEL }
}

export function isDiscoveryCustomOption(label: string) {
  return label.trim() === DISCOVERY_CUSTOM_OPTION_LABEL
}
